src/LayerEditor/ui/panels/surfaces.py

```diff
@@ -1,181 +1,78 @@
-        self._elevation = self.approximation.center()[2]
-
-
-    def set_name(self, name):
-        self.name = name
-        self._changed_forms = True
-
-    def set_name_from_file(self, names):
-        init_name = os.path.basename(self.grid_file)
-        init_name, _ext = os.path.splitext(init_name)
-        self.name = self.get_unique_name(init_name, names)
-
-    @staticmethod
-    def get_unique_name(init_name, names):
-        """
-        # TODO: make this a general function to provide unique default name. have general machanism to this in common. Given a list of names
-        # and given a name prefix return a first unique name.
-        # usage: get_unique_name(name, [surf.name for surf in surfaces.surfaces])
-
-        :param init_name:
-        :param names:
-        :return:
-        """
-        name_set = set(names)
-        namebase = name = init_name
-        i = 1
-
-        while name in name_set:
-            name = namebase + "_" + str(i)
-            i += 1
-        return name
-
-
-    def set_nuv(self, nuv):
-        self.nuv = nuv
-        self._changed_forms = True
-
-    def set_xy_transform(self, xy_transform):
-        self.xy_transform = xy_transform.tolist()
-        self._changed_forms = True
-
-    def get_actual_quad(self):
-        if self._quad is None:
-            return None
-        quad_center = np.average(self._quad, axis=0)
-        xy_mat = np.array(self.xy_transform, dtype = float)
-        return np.dot((self._quad - quad_center), xy_mat[0:2, 0:2].T) + quad_center + xy_mat[0:2, 2]
-
-    def compute_approximation(self):
-        self.approximation = self._approx_maker.compute_adaptive_approximation(
-            quad = self._quad,
-            nuv = self.nuv,
-            adapt_type="std_dev",
-            std_dev=self.tolerance)
-        self.update_from_z_surf()
-        self.approx_error = self._approx_maker.error
-        self._changed_forms = True
-
-    def save_to_layers(self, layers):
-        assert self.name != ""
-        assert self.grid_file != ""
-        assert self.approximation is not None
-        new_surface = copy.copy(self)
-        if self._idx == -1:
-            # New surface.
-            layers.add_surface(new_surface)
-            self._idx = len(layers.surfaces) - 1
-        else:
-            # Modify existing
-            layers.set_surface(self._idx, new_surface)
-        self._changed_forms = False
-
-
 class SurfacesComboBox(QtWidgets.QComboBox):
-    def __init__(self):
+    def __init__(self, parent):
         super().__init__()
+        self._parent = parent
         self.setEditable(True)
         self.setEnabled(False)
+        self.setLineEdit(LineEdit())
+        self.setCompleter(None)
+        self.lineEdit().textEdited.connect(self.text_changed)
+        self.lineEdit().editingFinished.connect(self.edit_finished)
+        self.setInsertPolicy(QtWidgets.QComboBox.NoInsert)
+        self.setValidator(TextValidator(self.unique_name_fnc,
```

Comment:
As the TextValidator is specific for usage here it would be better to move the code here, and simplify the function injection.

Reply:
It was meant to be general and possibly used in other similar cases. But I didnt got to it yet